src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,746 @@
+
+        let mut total_read = 0_usize;
+        let start_addr = usize::try_from(start_addr).expect("Unable to convert addr to usize");
+
+        while total_read < len {
+            let gaddr = start_addr + total_read;
+            let paddr =
+                match self.translate_gva(&vmm.vcpus_handles[tid_to_cpuid(tid)], gaddr as u64) {
+                    Ok(paddr) => usize::try_from(paddr).expect("Unable to convert addr to usize"),
+                    Err(e) => {
+                        error!("Error {e:?} translating gva on read address: {start_addr:X}");
+                        return Err(TargetError::NonFatal);
+                    }
+                };
+
+            let psize = arch::PAGE_SIZE;
+            let read_len = std::cmp::min(len - total_read, psize - (paddr & (psize - 1)));
+
+            if memory
+                .read(
+                    &mut data[total_read..total_read + read_len],
+                    GuestAddress(paddr as u64),
+                )
+                .is_err()
+            {
+                return Err(TargetError::NonFatal);
+            }
+
+            total_read += read_len;
+        }
+
+        Ok(total_read)
+    }
+
+    fn write_addrs(
+        &mut self,
+        start_addr: <Self::Arch as Arch>::Usize,
+        data: &[u8],
+        tid: Tid,
+    ) -> TargetResult<(), Self> {
+        let vmm = &self.vmm.lock().expect("Error unlocking vmm");
+        let memory = vmm.guest_memory();
+        let len = data.len();
+
+        let mut total_written = 0_usize;
+        let start_addr = usize::try_from(start_addr).expect("Unable to convert addr to usize");
+
+        while total_written < len {
+            let gaddr = start_addr + total_written;
+            let paddr =
+                match self.translate_gva(&vmm.vcpus_handles[tid_to_cpuid(tid)], gaddr as u64) {
+                    Ok(paddr) if paddr == <Self::Arch as Arch>::Usize::MIN => gaddr,
+                    Ok(paddr) => usize::try_from(paddr).expect("Error converting addr to usize"),
+                    Err(e) => {
+                        error!("Error {e:?} translating gva");
+                        return Err(TargetError::NonFatal);
+                    }
+                };
+
+            let psize = 4096;
+            let write_len = std::cmp::min(len - total_written, psize - (paddr & (psize - 1)));
+            if let Err(e) = memory.write(
+                &data[total_written..total_written + write_len],
+                GuestAddress(paddr as u64),
+            ) {
+                error!("Error {e:?} writing memory at {paddr:X}");
+                return Err(TargetError::NonFatal);
+            }
+            total_written += write_len;
+        }
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn list_active_threads(
+        &mut self,
+        thread_is_active: &mut dyn FnMut(Tid),
+    ) -> Result<(), Self::Error> {
+        let vmm = &self.vmm.lock().expect("Error unlocking vmm");
```

Comment:
Let's not lock the vmm mutex just to compute the number of vcpus. This can just as well be `self.vcpu_state.len()`, right?

Reply:
Fixed thanks